### PR TITLE
#60 Replay training dataset query

### PR DESCRIPTION
**Pull request by SirOibaf**

### CODE REVIEW

File: java/src/main/java/com/logicalclocks/hsfs/TrainingDataset.java [hunk 1]
```diff
@@ -157,11 +161,8 @@ public void save(Dataset<Row> dataset) throws FeatureStoreException, IOException
    * @throws IOException
    */
   public void save(Query query, Map<String, String> writeOptions) throws FeatureStoreException, IOException {
-    Dataset<Row> dataset = query.read();
-    trainingDatasetEngine.save(this, dataset, writeOptions);
-    if (statisticsEnabled) {
-      statisticsEngine.computeStatistics(this, dataset);
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/TrainingDataset.java:163]: why did you remove the statistics computation?

File: java/src/main/java/com/logicalclocks/hsfs/engine/TrainingDatasetEngine.java [hunk 2]
```diff
@@ -65,6 +69,7 @@ public void save(TrainingDataset trainingDataset, Dataset<Row> dataset,
     }
 
     // Update the original object - Hopsworks returns the full location and incremented version
+    trainingDataset.setId(apiTD.getId());
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/engine/TrainingDatasetEngine.java:72]: duplicate of line 75

File: java/src/main/java/com/logicalclocks/hsfs/engine/Utils.java [hunk 3]
```diff
@@ -48,10 +49,22 @@
...
+    int index = 0;
+    for (StructField structField : dataset.schema().fields()) {
+      // TODO(Fabio): unit test this one for complext types
+      features.add(new TrainingDatasetFeature(structField.name(), structField.dataType().catalogString(), index++));
+    }
+
+    return features;
+  }
+
+  public void schemaMatches(Dataset<Row> dataset, List<TrainingDatasetFeature> features) throws FeatureStoreException {
```
Comment (moritzmeister) [inline java/src/main/java/com/logicalclocks/hsfs/engine/Utils.java:64]: maybe we can call it `trainingDatasetSchemaMatches` to avoid confusion
```suggestion
  public void trainingDatasetSchemaMatches(Dataset<Row> dataset, List<TrainingDatasetFeature> features) throws FeatureStoreException {
```

File: python/hsfs/core/training_dataset_engine.py [hunk 4]
```diff
@@ -72,6 +72,11 @@ def read(self, training_dataset, split, user_read_options):
             path,
         )
 
+    def query(self, training_dataset, storage):
+        return self._training_dataset_api.get_query(training_dataset)[
+            "queryOnline" if storage == "online" else "query"
```
Comment (moritzmeister) [inline python/hsfs/core/training_dataset_engine.py:77]: should probably `.lower()` the storage

File: python/hsfs/feature_group.py [hunk 5]
```diff
@@ -67,15 +67,6 @@ def __init__(
         self._default_storage = default_storage
         self._hudi_enabled = hudi_enabled
 
-        if id is None:
```
Comment (moritzmeister) [inline python/hsfs/feature_group.py:70]: am I stupid, why did I add this twice, thanks for fixing

File: python/hsfs/training_dataset.py [hunk 6]
```diff
@@ -409,3 +419,6 @@ def get_statistics(self, commit_time=None):
             return self.statistics
         else:
             return self._statistics_engine.get(self, commit_time)
+
+    def query(self, storage="online"):
```
Comment (moritzmeister) [inline python/hsfs/training_dataset.py:423]: from an api perspective, I think this could be a property, and then have a separate method `.get_query(storage)` see the statistics, how I did it there. not sure what's better.

**Review by moritzmeister — CHANGES_REQUESTED** [review]
I wasn't able to add a comment there because you didn't touch that part of the file but in `spark.py` engine, I think we need to fix the `schema_matches` function to take into account the order since we use it for inserting data into training datasets. The same as you did in the java part.

File: java/src/main/java/com/logicalclocks/hsfs/TrainingDataset.java (hunk 1 above)
Comment (SirOibaf) [reply java/src/main/java/com/logicalclocks/hsfs/TrainingDataset.java:163]: No I didn't. `save(Query query... ` calls `save(Dataset<Row>` which has the code to compute the statistics

File: java/src/main/java/com/logicalclocks/hsfs/engine/TrainingDatasetEngine.java (hunk 2 above)
Comment (SirOibaf) [reply java/src/main/java/com/logicalclocks/hsfs/engine/TrainingDatasetEngine.java:72]: done

File: java/src/main/java/com/logicalclocks/hsfs/engine/Utils.java (hunk 3 above)
Comment (SirOibaf) [reply java/src/main/java/com/logicalclocks/hsfs/engine/Utils.java:64]: Ok

File: java/src/main/java/com/logicalclocks/hsfs/TrainingDataset.java (hunk 1 above)
Comment (moritzmeister) [reply java/src/main/java/com/logicalclocks/hsfs/TrainingDataset.java:163]: ah yes, sorry

File: python/hsfs/core/training_dataset_engine.py (hunk 4 above)
Comment (SirOibaf) [reply python/hsfs/core/training_dataset_engine.py:77]: fixed

File: python/hsfs/training_dataset.py (hunk 6 above)
Comment (SirOibaf) [reply python/hsfs/training_dataset.py:423]: Ok, I see, let's keep it consistent then.

Comment (SirOibaf) [reply python/hsfs/training_dataset.py:423]: fixed and added docstring

**Review by moritzmeister — APPROVED** [review]
:+1:
